Code/CryGame/Items/Weapons/OffHand.cpp

```diff
@@ -1516,43 +1544,88 @@ void COffHand::NetStartFire()
 
 	CWeapon::NetStartFire();
 
-	AttachGrenadeToHand(GetCurrentFireMode(), false, true);
+	if (!gEnv->bClient)
+		return;
+
+	if (m_heldEntityId)
+	{
+		CPlayer* pPlayer = CPlayer::FromActor(GetOwnerActor());
+		if (pPlayer)
+		{
+			pPlayer->StartThrowPrep();
+		}
+	}
+	else
+	{
+		AttachGrenadeToHand(GetCurrentFireMode(), m_stats.fp, true);
+	}
 
 	//Handle FP Spectator
-	if (m_stats.fp)
+	if (m_stats.fp && GetOwnerActor())
 	{
-		m_mainHand = static_cast<CItem*>(GetOwnerActor()->GetCurrentItem());
-		if (m_mainHand)
+		CWeapon *pMainWeapon = GetOwnerActor()->GetCurrentWeapon(false);
+		SetMainHandWeapon(pMainWeapon);
+		
+		if (pMainWeapon)
 		{
 			//if (!(m_currentState & (eOHS_THROWING_NPC | eOHS_THROWING_OBJECT)))
 			{
-				if (m_mainHandWeapon && m_mainHandWeapon->IsWeaponRaised())
+				if (pMainWeapon->IsWeaponRaised())
 				{
-					m_mainHandWeapon->RaiseWeapon(false, true);
-					m_mainHandWeapon->SetDefaultIdleAnimation(CItem::eIGS_FirstPerson, g_pItemStrings->idle);
+					pMainWeapon->RaiseWeapon(false, true);
+					pMainWeapon->SetDefaultIdleAnimation(CItem::eIGS_FirstPerson, g_pItemStrings->idle);
 				}
-				m_mainHand->PlayAction(g_pItemStrings->offhand_on);
-				m_mainHand->SetActionSuffix("akimbo_");
+				pMainWeapon->PlayAction(g_pItemStrings->offhand_on);
+				pMainWeapon->SetActionSuffix("akimbo_");
 			}
 		}
-		if (m_mainHandWeapon && m_mainHandWeapon->GetEntity()->GetClass() == CItem::sFistsClass)
+		if (pMainWeapon && pMainWeapon->GetEntity()->GetClass() == CItem::sFistsClass)
 		{
-			CFists* pFists = static_cast<CFists*>(m_mainHandWeapon);
+			CFists* pFists = static_cast<CFists*>(pMainWeapon);
 			pFists->RequestAnimState(CFists::eFAS_FIGHT);
 		}
 	}
 }
 
+//=============================================================================
 void COffHand::NetStopFire()
 {
 	CWeapon::NetStopFire();
 
+	if (!gEnv->bClient)
+	{
+		CActor* pOwner = GetOwnerActor();
+		if (pOwner && pOwner->GetHeldObjectId())
+		{
+			pOwner->OnObjectEvent(CActor::ObjectEvent::THROW, pOwner->GetHeldObjectId());
+		}
+		return;
+	}
+
 	AttachGrenadeToHand(GetCurrentFireMode(), m_stats.fp, false);
 
-	//Handle FP Spectator
-	if (m_stats.fp)
+	if (m_heldEntityId)
+	{
+		CPlayer* pPlayer = CPlayer::FromActor(GetOwnerActor());
+		if (pPlayer)
+		{
+			pPlayer->OnObjectEvent(CActor::ObjectEvent::THROW, m_heldEntityId);
+		}
+
+		if (GetMainHandWeapon() && !GetMainHandWeapon()->IsDualWield())
+		{
+			GetMainHandWeapon()->PlayAction(g_pItemStrings->offhand_off, 0, false, CItem::eIPAF_Default | CItem::eIPAF_NoBlend);
+		}
+	}
+	else
 	{
-		GetScheduler()->TimerAction(GetCurrentAnimationTime(CItem::eIGS_FirstPerson), CSchedulerAction<FinishGrenadeAction>::Create(FinishGrenadeAction(this, m_mainHand)), false);
+		GetScheduler()->TimerAction(
+			GetCurrentAnimationTime(CItem::eIGS_FirstPerson),
+			MakeAction([this, hand = GetMainHandWeapon()](CItem*) {
```

Comment:
Are we sure the `hand` pointer will always be valid when the lambda is called? There are two more cases below. Also, what about `this`? Is there a chance that the `COffhand` object gets destroyed before the lambda is called?